stm32-modules/heater-shaker/include/heater-shaker/host_comms_task.hpp

```diff
@@ -39,9 +60,306 @@ requires MessageQueue<QueueImpl<Message>, Message> class HostCommsTask {
+        static_cast<void>(tx_into);
+        static_cast<void>(tx_length_available);
+        return std::make_pair(true, 0);
+    }
+
+    auto visit_gcode(const gcode::SetRPM& gcode, char* tx_into,
+                     size_t tx_length_available) -> std::pair<bool, size_t> {
+        auto id = ack_only_cache.add(gcode);
+        if (id == 0) {
+            return std::make_pair(false,
+                                  errors::write_into<errors::GCodeCacheFull>(
+                                      tx_into, tx_length_available));
+        }
+        auto message =
+            messages::SetRPMMessage{.id = id, .target_rpm = gcode.rpm};
+        if (!task_registry->motor.get_message_queue().try_send(
+                message, TICKS_TO_WAIT_ON_SEND)) {
+            auto written = errors::write_into<errors::InternalQueueFull>(
+                tx_into, tx_length_available);
+            ack_only_cache.remove_if_present(id);
+            return std::make_pair(false, written);
+        }
+        return std::make_pair(true, 0);
+    }
+
+    auto visit_gcode(const gcode::GetRPM& gcode, char* tx_into,
+                     size_t tx_length_available) -> std::pair<bool, size_t> {
+        auto id = get_rpm_cache.add(gcode);
+        if (id == 0) {
+            return std::make_pair(false,
+                                  errors::write_into<errors::GCodeCacheFull>(
+                                      tx_into, tx_length_available));
+        }
+        auto message = messages::GetRPMMessage{.id = id};
+        if (!task_registry->motor.get_message_queue().try_send(
+                message, TICKS_TO_WAIT_ON_SEND)) {
+            auto written = errors::write_into<errors::InternalQueueFull>(
+                tx_into, tx_length_available);
+            get_rpm_cache.remove_if_present(id);
+            return std::make_pair(false, written);
+        }
+        return std::make_pair(true, 0);
+    }
+
+    auto visit_gcode(const gcode::SetTemperature& gcode, char* tx_into,
+                     size_t tx_length_available) -> std::pair<bool, size_t> {
+        auto id = ack_only_cache.add(gcode);
+        if (id == 0) {
+            return std::make_pair(false,
+                                  errors::write_into<errors::GCodeCacheFull>(
+                                      tx_into, tx_length_available));
+        }
+        auto message = messages::SetTemperatureMessage{
+            .id = id,
+            .target_temperature = static_cast<uint32_t>(gcode.temperature)};
+        if (!task_registry->heater.get_message_queue().try_send(
+                message, TICKS_TO_WAIT_ON_SEND)) {
+            auto written = errors::write_into<errors::InternalQueueFull>(
+                tx_into, tx_length_available);
+            ack_only_cache.remove_if_present(id);
+            return std::make_pair(false, written);
+        }
+        return std::make_pair(true, 0);
+    }
+
+    auto visit_gcode(const gcode::GetTemperature& gcode, char* tx_into,
```

Comment:
cool use of overloads

Reply:
yeah. imo it's kind of clunky syntax wise, i'd much prefer a match-statement syntax like [rust](https://doc.rust-lang.org/rust-by-example/flow_control/match.html) but c++ is very conservative about that sort of thing